tests/testapp/test_queries.py

```diff
@@ -753,3 +753,172 @@ def test_order_by_related(self):
                 tree.child2_2,
             ],
         )
+
+    def test_tree_exclude(self):
+        tree = self.create_tree()
+        # Tree-filter should remove children if
+        # the parent meets the filtering criteria
+        nodes = Model.objects.tree_exclude(name="2")
+        self.assertEqual(
+            list(nodes),
+            [
+                tree.root,
+                tree.child1,
+                tree.child1_1,
+            ],
+        )
+
+    def test_tree_filter(self):
+        tree = self.create_tree()
+        # Tree-filter should remove children if
+        # the parent does not meet the filtering criteria
+        nodes = Model.objects.tree_filter(name__in=["root","1-1","2","2-1","2-2"])
+        self.assertEqual(
+            list(nodes),
+            [
+                tree.root,
+                tree.child2,
+                tree.child2_1,
+                tree.child2_2,
+            ],
+        )
+
+    def test_tree_filter_chaining(self):
+        tree = self.create_tree()
+        # Tree-filter should remove children if
+        # the parent does not meet the filtering criteria
+        nodes = Model.objects.tree_exclude(name="2-2").tree_filter(name__in=["root","1-1","2","2-1","2-2"])
+        self.assertEqual(
+            list(nodes),
+            [
+                tree.root,
+                tree.child2,
+                tree.child2_1,
+            ],
+        )
+
+    def test_tree_filter_related(self):
+        tree = type("Namespace", (), {})()  # SimpleNamespace for PY2...
```

Comment:
We do not support Python 2 anymore, maybe use `types.SimpleNamespace` directly?

Reply:
Would you like me to change over the other namespaces that are in the test suite while I'm in there?

Reply:
Yes, that would be great. Thank you!

Reply:
Alright, I've swapped all the namespaces in the test suite over.